client/clienterror/clienterror.go

```diff
@@ -0,0 +1,173 @@
+	if ok {
+		return &APIError{
+			HTTPStatusCode: http.StatusUnauthorized,
+			OriginalError:  convertedError,
+			ErrorMessage:   "Unauthorized",
+			ErrorDetails:   "The token presented was likely no longer valid. No further action is required.",
+		}
+	}
+
+	// HTTP level error cases
+	runtimeAPIError, runtimeAPIErrorOK := err.(*runtime.APIError)
+	if runtimeAPIErrorOK {
+		ae := &APIError{
+			HTTPStatusCode: runtimeAPIError.Code,
+			OriginalError:  runtimeAPIError,
+			ErrorMessage:   fmt.Sprintf("HTTP Error %d", runtimeAPIError.Code),
+		}
+
+		// special messages for some codes
+		switch runtimeAPIError.Code {
+
+		case http.StatusForbidden:
+			ae.ErrorMessage = "Forbidden"
+			ae.ErrorDetails = "The client has been denied access to the API endpoint with an HTTP status of 403.\n"
+			ae.ErrorDetails += "Please make sure that you are in the right network or VPN. Once that is verified,\n"
+			ae.ErrorDetails += "check back with Giant Swarm support that your network is permitted access."
+
+		case http.StatusUnauthorized:
+			ae.ErrorMessage = "Not authorized"
+			ae.ErrorDetails = "You are not authorized for this action.\n"
+			ae.ErrorDetails += "Please check whether you are logged in with the given endpoint."
+
+		case http.StatusInternalServerError:
+			ae.ErrorMessage = "Backend error"
+			ae.ErrorDetails = "The backend responded with an HTTP 500 code, indicating an internal error on Giant Swarm's side.\n"
+			ae.ErrorDetails += "Original error message: " + runtimeAPIError.Error() + "\n"
+			ae.ErrorDetails += "Please report this problem to the Giant Swarm support team."
+		}
+
+		return ae
+	}
+
+	// Errors on levels lower than HTTP
+	// is url.Error
+	urlError, urlErrorOK := err.(*url.Error)
+	if urlErrorOK {
+		ae := &APIError{
+			OriginalError: urlError,
+			URL:           urlError.URL,
+			HTTPMethod:    urlError.Op,
+		}
+
+		// is net.OpError
+		netOpError, netOpErrorOK := urlError.Err.(*net.OpError)
+		if netOpErrorOK {
+			ae.OriginalError = netOpError
+
+			// is net.DNSError
+			netDNSError, netDNSErrorOK := netOpError.Err.(*net.DNSError)
+			if netDNSErrorOK {
+				ae.OriginalError = netDNSError
+				ae.IsTemporary = netDNSError.IsTemporary
+				ae.IsTimeout = netDNSError.IsTimeout
+				ae.ErrorMessage = "DNS error"
+				ae.ErrorDetails = fmt.Sprintf("The host name '%s' cannot be resolved.\n", netDNSError.Name)
+				ae.ErrorDetails += "Please make sure the endpoint URL you are using is correct."
+
+				return ae
+			}
+
+			// dial error
+			if netOpError.Op == "dial" {
+				ae.ErrorMessage = "No connection to host"
+				ae.ErrorDetails = fmt.Sprintf("The host '%s' cannot be reached.\n", netOpError.Addr.String())
+				ae.ErrorDetails += "Please make sure that you are in the right network or VPN."
+
+				return ae
+			}
+		}
+
+		return ae
+	}
+
+	// Timeout / context deadline exceeded
+	// Note: We'd love to do type assertion here, but there is no exported type to assert with.
+	errorType := reflect.TypeOf(err)
```

Comment:
I would rather have `err == context.DeadlineExceeded`

Reply:
Cool! I missed that.